### ISSUE TITLE
Suggestion: Change scaling of the max_distance levels in visualize_hierarchical_documents

### ISSUE THREAD

**Issue by zilch42** [MaartenGr/BERTopic#1077]
Hi Maarten,

The `visualize_hierarchical_documents` function is super cool for visualising the topic structure so thank you for your work on that. 

As a suggestion, I wonder if it would be helpful to have some control over the resolution at different ends of the hierarchy? Once the number of topics becomes reasonably large (>20ish) the plot becomes less helpful because the labels dominate and the colours get reused repeatedly. With very large numbers of topics (>100), even with a higher `nr_levels`, most of those levels aren't very useful because they are just a soup of hundreds of topics. What would be great is having less resolution at the lower end of the slider and more resolution at the upper end of the slider where there are fewer higher-level topics, as it's quite interesting to see how those higher-level topics split out and merge. 

![image](https://user-images.githubusercontent.com/39004650/223337564-821390b7-9d89-442c-9463-7c757a1a2ebf.png)

Splitting the distance indices over a logarithmic scale rather than a linear one seems to produce something more useful to me.

I'm sure there is a more elegant way to do this but as an example, replacing this line with the code below produced something more usable for me. 
https://github.com/MaartenGr/BERTopic/blob/d665d3f8d8c7c1736dc82b1df8839ced56a2adb6/bertopic/plotting/_hierarchical_documents.py#L150

```
    log_indices = np.round(np.logspace(start=math.log(1,10), stop=math.log(len(distances)-1,10), num=nr_levels)).astype(int).tolist()
    log_indices.reverse()
    max_distances = [distances[i] for i in log_indices]
```
Ideally the user should probably have some control over the scaling too.

Open to your thoughts.

**Comment by MaartenGr** [comment]
Thanks for the suggestion! Could you perhaps create a reproducible example of this? Or perhaps show the effect of what you propose? In practice, I wonder what the effect would be of your suggestion on use cases where hierarchical topic modeling is applied on < 30 topics with > 5 levels.

**Comment by zilch42** [comment]
Sure thing. Forgive me, I am not 100% sure on the best way to provide a reproducible example that relies on changes to your source code. Is a pull request the easiest way to do that? Anyway, the following example assumes the following changes have been made to BERTopic:

`log_scale_levels: bool = False` added as an input argument in 2 places :
https://github.com/MaartenGr/BERTopic/blob/d665d3f8d8c7c1736dc82b1df8839ced56a2adb6/bertopic/plotting/_hierarchical_documents.py#L9

https://github.com/MaartenGr/BERTopic/blob/7d45cf710aefff178d54403989e90432e0c49910/bertopic/_bertopic.py#L2145
 
Defaulting to `False` keeps the current behaviour. 

----------------------
`log_scale_levels=log_scale_levels` added to the call here:
https://github.com/MaartenGr/BERTopic/blob/7d45cf710aefff178d54403989e90432e0c49910/bertopic/_bertopic.py#L2239

  -----------------------
The following block of code:
```
    if log_scale_levels:
        log_indices = np.round(np.logspace(start=math.log(1,10), stop=math.log(len(distances)-1,10), num=nr_levels)).astype(int).tolist()
        log_indices.reverse()
        max_distances = [distances[i] for i in log_indices]
    else:
        max_distances = [distances[indices[-1]] for indices in np.array_split(range(len(hierarchical_topics)), nr_levels)][::-1]
```
replacing this line:

https://github.com/MaartenGr/BERTopic/blob/d665d3f8d8c7c1736dc82b1df8839ced56a2adb6/bertopic/plotting/_hierarchical_documents.py#L150 
------------------------------

# Example code (400 topics):
It's worth saying before all this that splitting the newsgroups data into 400 clusters and then merging them doesn't look great. it produces a bit of nonsense, but its available and illustrates the purpose. I do get more reasonable results with some of my own data which is naturally hierarchical but unfortunately can't share it.

It also might be more helpful to run these and play with the slider yourself.

```
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups
from sentence_transformers import SentenceTransformer

docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = embedding_model.encode(docs, show_progress_bar=True)

topic_model = BERTopic(embedding_model=embedding_model, nr_topics=400)
topics, probs = topic_model.fit_transform(docs, embeddings)
hierarchical_topics = topic_model.hierarchical_topics(docs)
```

## Linear Level 0
All topics is obviously a mess 
`topic_model.visualize_hierarchical_documents(docs, hierarchical_topics, embeddings=embeddings, log_scale_levels=False)`

![image](https://user-images.githubusercontent.com/39004650/224193972-83933ed7-37ce-488f-8c9a-5de529ce51fe.png)

## Linear Level 6
By level 6 we can start to pick out individual topics but there aren't many levels left to go 
![image](https://user-images.githubusercontent.com/39004650/224194098-75f7a580-c2e5-47fd-bb76-52edaa598e8f.png)

## Linear Level 9
At the top level there are still quite a few topics and we would probably like to go further.
![image](https://user-images.githubusercontent.com/39004650/224194229-f4f5f3d0-3c5f-4f86-8e0f-23b283134057.png)

## Logarithmic Level 3
Now let's run with the logarithmic scale.
`topic_model.visualize_hierarchical_documents(docs, hierarchical_topics, embeddings=embeddings, log_scale_levels=True)`

Level 0 is the same as linear so is not shown, but by Level 3 we start to have few enough topics that we can actually distinguish between them (~ equivalent to linear level 6)
![image](https://user-images.githubusercontent.com/39004650/224197419-552531d6-1e2d-46e5-9d11-3f0bff398d7b.png)

## Logarithmic Level 6
At level 6 we only have a handful of topics and we can start to see the effect of just one or two major topics merging by going further.
![image](https://user-images.githubusercontent.com/39004650/224194726-07ba05a2-2f69-47eb-9586-a524313a9ecd.png)

## Logarithmic Level 9
By level 9 we actually get to the top of the tree (2 topics left)
![image](https://user-images.githubusercontent.com/39004650/224197191-4cd99e20-561a-4f0b-a2cb-8038e0252989.png)

-----------------------------------

# Example code (20 topics):

```
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups
from sentence_transformers import SentenceTransformer

docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

embedding_model = SentenceTransformer("all-MiniLM-L6-v2")
embeddings = embedding_model.encode(docs, show_progress_bar=True)

topic_model = BERTopic(embedding_model=embedding_model, nr_topics=20)
topics, probs = topic_model.fit_transform(docs, embeddings)
hierarchical_topics = topic_model.hierarchical_topics(docs)
```

## Level 0
Linear:
`topic_model.visualize_hierarchical_documents(docs, hierarchical_topics, embeddings=embeddings, log_scale_levels=False)`
Logarithmic:
`topic_model.visualize_hierarchical_documents(docs, hierarchical_topics, embeddings=embeddings, log_scale_levels=True)`
![image](https://user-images.githubusercontent.com/39004650/224198793-2f0c1152-a23c-497b-bc0a-982c5b4af0a4.png)

I've only shown Level 0 here. You can run and play with these yourself if you like. Level 0 with only twenty topics is already quite manageable and linear splitting of the levels reduces two clusters each level which is great. Logarithmic splitting with low numbers of clusters can result in some levels being the same e.g. level 8 and 9 in this case are the same with logarithmic scaling. You could do some magic and check the logarithmic vector for levels that were identical and nudge them so that you are at least merging one cluster each level, but I suggest just having a linear/logarithmic argument like `log_scale_levels` and letting the user choose which is most appropriate for their situation.

**Comment by MaartenGr** [comment]
Thanks for the extensive examples! It indeed seems quite a bit nicer when the number of topics is rather large. The effect of a logarithmic scale seems to be quite impactful. My only concern is that linear might work better at a lower number of topics but that can be up to the user, as long as `log_scale_levels` is set to False as a default. If you want, I would appreciate a PR implementing this. Perhaps renaming the parameter to something like `level_scale` might fit a bit better with its corresponding `nr_levels` parameter.
